qiskit_ibm_provider/transpiler/passes/scheduling/block_base_padder.py

```diff
@@ -79,57 +102,69 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
-            if node in self._node_start_time:
-                if isinstance(node.op, Delay):
-                    self._visit_delay(node)
-                else:
-                    self._visit_generic(node)
+        # Top-level dag is the entry block
+        new_dag = self._visit_block(dag)
 
-            else:
-                raise TranspilerError(
-                    f"Operation {repr(node)} is likely added after the circuit is scheduled. "
-                    "Schedule the circuit again if you transformed it."
-                )
-
-            self._prev_node = node
-
-        # terminate final block
-        self._terminate_block(self._block_duration, self._current_block_idx, None)
-
-        return self._dag
+        return new_dag
 
     def _init_run(self, dag: DAGCircuit) -> None:
         """Setup for initial run."""
         self._node_start_time = self.property_set["node_start_time"].copy()
+        self._node_block_dags = self.property_set["node_block_dags"]
         self._idle_after = {bit: 0 for bit in dag.qubits}
         self._current_block_idx = 0
         self._conditional_block = False
         self._block_duration = 0
 
         # Prepare DAG to pad
-        self._dag = DAGCircuit()
-        for qreg in dag.qregs.values():
-            self._dag.add_qreg(qreg)
-        for creg in dag.cregs.values():
-            self._dag.add_creg(creg)
-
-        # Update start time dictionary for the new_dag.
-        # This information may be used for further scheduling tasks,
-        # but this is immediately invalidated because most node ids are updated in the new_dag.
+        self._root_dag = dag
+        self._dag = self._empty_dag_like(dag)
+        self._block_dag = self._dag
+        self._bit_indices = {q: index for index, q in enumerate(dag.qubits)}
+        self._last_node_to_touch = {}
+        self._fast_path_nodes = set()
+        self._dirty_qubits = set()
+
         self.property_set["node_start_time"].clear()
+        self._prev_node = None
 
-        self._dag.name = dag.name
-        self._dag.metadata = dag.metadata
-        self._dag.unit = self.property_set["time_unit"]
-        self._dag.calibrations = dag.calibrations
-        self._dag.global_phase = dag.global_phase
+    def _empty_dag_like(self, dag: DAGCircuit, pad_wires: bool = True) -> DAGCircuit:
+        """Create an empty dag like the input dag."""
+        new_dag = DAGCircuit()
 
-        self._prev_node = None
+        # Ensure *all* registers are included from the input circuit
+        # so that they are scheduled in sub-blocks
+
+        # Control flow blocks do not get the full reg added to the
+        # block but just the bits. To work around this we try to
+        # add the reg if available and otherwise add the bits directly.
+        # We need this work around as otherwise the padded circuit will
+        # not be equivalent to one written manually as bits will not
+        # be defined on registers like in the test case.
```

Comment:
I tried this approach initially but unfortunately, it does matter from the perspective of equivalency testing of the padded circuits compared with the handwritten circuits using the control flow builder.

Basically the problem is that the "top-level" circuit has the registers for equivalency testing, but the nested blocks don't. This method handles that case.

In general, I find having a "top-level" circuit was triggering many edge cases that might be mitigated with the approach of MLIR where the top-level "module" itself is an operations containing blocks containing operations. This would move the role of a QuantumCircuit to be a block and would require introducing a top-level "CircuitModule" equivalent.

Reply:
Updated the comment here for clarity 2144c88.